Scene publishing support

keianhzo opened the pull request:
This PR adds support for scene publishing

Exairnous commented:
Thanks!  I'll review this as soon as I can.

Exairnous reviewed (changes requested):
My apologies for the long delay in reviewing this.  I think this will be a fantastic addition, but I have some notes:
* The attribution field and remixing/promotion checkboxes are missing.
* The json module is needed several times in the api.py file, so I think it might as well be imported at the top of the file.
* Why were the context pointers, etc. changed in environment_settings.py and operators.py?  They seem unrelated.
* I talked to @j-conrad at one point and he mentioned that he had added the old update scene operator to the quick favorites menu and now it updates the published scene rather than the local scene in the room.  I know this is because the idnames were updated to be more descriptive, but it might be nice if they could be tweaked slightly so this could be avoided.
* The warning for needing to be signed in should only be applied when updating the scene, not spawning objects, also it looks like my fix for allowing spawned objects hasn't been merged in (although, my fix may need to be tweaked because of the new requirements of sign in for publishing/updating scenes).
* I'm not sure the annotations you added are needed.  Currently they're applied inconsistently and they don't provide any additional information in my opinion.
* I don't think the scene list should be cached and/or saved to the json prefs.  This adds a lot of complexity just to allow a speed boost in getting scenes for an instance when reloaded, can easily be outdated, and can cause poll issues with the other operators in that section when switching instances (because the items are technically still there, just hidden).  Unless it's really slow getting a long list of scenes (once the unneeded export scene is removed), I don't think the caching is worth it. Also, I think the scene list should be cleared when the session is closed.
* Underscores shouldn't be allowed in the scene name and a warning should be displayed.  A warning should also be shown for an empty scene name.
* I don't think there's any way to update the screenshot/room name/attribution/remixing/promotion of an already published scene.  It would be nice if there were a checkbox or something to allow this.
* The update scene confirmation doesn't provide very good user feedback in Blender 4.0 (the menu doesn't disappear after clicking on yes for me).  I'm not sure if there's anything to do about this or not, but I thought I should mention it.

keianhzo commented on addons/io_hubs_addon/api.py:
```diff
@@ -0,0 +1,92 @@
+def upload_media(endpoint, file):
+    resp = requests.post(f'{endpoint}/api/v1/media', files={'media': (
+        'glb', file, 'application/octet-stream')}, verify=False)
+    json = resp.json()
```
Updated.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -297,28 +291,36 @@ def draw(self, context: Context):
+            row = box.row()
+            row.alert = True
+            row.label(
+                text="You need to be signed in Hubs to update the room scene or spawn objects")
```
Thanks updated.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -525,43 +551,315 @@ def draw_item(self, context, layout, data, item, icon, active_data, active_propn
+                "screenshot_file_id": screenshot_data["file_id"],
+                "screenshot_file_token": screenshot_data["access_token"]
+            })
+            print(screenshot_data)
```
Removed

keianhzo replied:
Removed.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -525,43 +551,315 @@ def draw_item(self, context, layout, data, item, icon, active_data, active_propn
+            screenshot = context.scene.hubs_scene_debugger_scene_publish_props.screenshot
+            screenshot_full = bpy.path.abspath(
+                screenshot.filepath, library=screenshot.library)
+            screenshot_norm = os.path.normpath(screenshot_full)
```
Honestly I don't remember, I might have run into some issues at the time. I don't think it harms.

keianhzo replied:
Not really, I think I've might read about it at the time.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -525,43 +551,315 @@ def draw_item(self, context, layout, data, item, icon, active_data, active_propn
+                "model_file_token": glb_data["access_token"]
+            })
+
+            screenshot = context.scene.hubs_scene_debugger_scene_publish_props.screenshot
```
Good idea, added.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -525,43 +551,315 @@ def draw_item(self, context, layout, data, item, icon, active_data, active_propn
+        return {'FINISHED'}
+
+
+class HubsCreateSceneOperator(bpy.types.Operator):
```
Yesy, good idea.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -622,6 +920,103 @@ class HubsSceneDebuggerRoomExportPrefs(bpy.types.PropertyGroup):
+        name="Id",
+        description="Scene id",
+        default="Scene id",
+        get=lambda self: self["scene_id"]
```
I think it was simple enough to be a good lambda case.

Exairnous replied:
But doesn't it just do the same as the default?  If so, why add it?

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -525,43 +551,315 @@ def draw_item(self, context, layout, data, item, icon, active_data, active_propn
+        scenes_props.instance = hubs_session.reticulum_url
+        scenes_props.scenes.clear()
+        try:
+            export_scene(context)
```
Oh good catch, that's bad, I don't know how that ended up there, probably I copy&pasted some code. Removed.

keianhzo commented:
Thanks a lot for the review. I think I have addressed quite some important issues. for the rest:

> The attribution field and remixing/promotion checkboxes are missing.
Yes, that's important, but doing it right will require some time and I would rather not be delaying this feature anymore as I think it can be very useful. Maybe some good Samaritan would be willing to add support for that. It shouldn't be difficult, it's mostly a matter of adding a list with "creator" and "url" that get added to the attributions object in scene_data when publishing or updating. We would probably need two lists UI widgets, one inside "manage" and another one inside "Publish".

> Why were the context pointers, etc. changed in environment_settings.py and operators.py? They seem unrelated.
We are using the `hubs_open_image operator` for the screenshot, and I needed to decouple it from components and use `getattr/setattr` to make it work with `context.scene`.

> I talked to @j-conrad at one point and he mentioned that he had added the old update scene operator to the quick favorites menu and now it updates the published scene rather than the local scene in the room. I know this is because the idnames were updated to be more descriptive, but it might be nice if they could be tweaked slightly so this could be avoided.
This shouldn't happen anymore, right? I updated the name as you described above, hope that helps.

> I don't think the scene list should be cached and/or saved to the json prefs. This adds a lot of complexity just to allow a speed boost in getting scenes for an instance when reloaded, can easily be outdated, and can cause poll issues with the other operators in that section when switching instances (because the items are technically still there, just hidden). Unless it's really slow getting a long list of scenes (once the unneeded export scene is removed), I don't think the caching is worth it. Also, I think the scene list should be cleared when the session is closed.
We can address this in a follow-up if it's a real problem, dealing with undoing that delay this release quite lot.

> Underscores shouldn't be allowed in the scene name and a warning should be displayed. A warning should also be shown for an empty scene name.
Another good follow-up.

> I don't think there's any way to update the screenshot/room name/attribution/remixing/promotion of an already published scene. It would be nice if there were a checkbox or something to allow this.
Related to my comment above.

> The update scene confirmation doesn't provide very good user feedback in Blender 4.0 (the menu doesn't disappear after clicking on yes for me). I'm not sure if there's anything to do about this or not, but I thought I should mention it.
Same, good follow up.

Exairnous commented:
> Thanks a lot for the review. 

You're welcome :)

> We are using the hubs_open_image operator for the screenshot, and I needed to decouple it from components and use getattr/setattr to make it work with context.scene.

Ah, okay.  Thanks.

> This shouldn't happen anymore, right? I updated the name as you described above, hope that helps.

This was an issue with the idname for the new `HubsUpdateSceneOperator` being the same as the old idname for the `HubsUpdateRoomOperator`.  I don't think you changed anything there, so if people had the old operator bound to the quick favorites menu it's probably still triggering the new operator now.  But it was only a small issue, I doubt many people will run into it and Jim already knows to re-add it.

> We can address this in a follow-up if it's a real problem, dealing with undoing that delay this release quite a lot.

Well, it introduces a number of corner cases, but sure, it can be updated in a followup PR.


I guess, if it's alright with you, I'll address the remaining things in followup PRs after the release.

keianhzo commented:
Yep, taht sounds great!
